### PR TITLE
freeze schemapack and datapack spec models (GSI-653)

### PR DESCRIPTION
**Pull request by KerstenBreuer**
```
- make datapack and schemapack spec models immutable
- allow hashing
- expand serialization tests

Bumps version to 2.0.0-alpha.4.
```

### CODE REVIEW

File: src/schemapack/_internals/validation/plugins/content_schema.py [hunk 1]
```diff
@@ -71,8 +72,9 @@ def validate(
         Raises:
             schemapack.exceptions.ValidationPluginError: If validation fails.
         """
+        json_compatible_content = json.loads(resource.model_dump_json())["content"]
```
Comment (lkuchenb) [inline src/schemapack/_internals/validation/plugins/content_schema.py:75]: Why is this necessary?

**Review by lkuchenb — DISMISSED** [review]
Only comments out of interest, see below. Also, why can't we work with the existing frozendict library?

File: src/schemapack/_internals/validation/plugins/content_schema.py (hunk 1 above)
Comment (KerstenBreuer) [reply src/schemapack/_internals/validation/plugins/content_schema.py:75]: The json standard library can only work with standard data types. Thus it does not know what FrozenDict is and you cannot make it aware of it. In deed it is doing isinstance checks directly on the commonly used standard types such as dict but not on generic types such as Mapping. The latter would have worked out of the box with FrozenDict. Thus we need to convert to standard dictionaries and the serialization option by pydantic is the easiest way to to so.

File: src/schemapack/_internals/spec/schemapack.py [hunk 2]
```diff
@@ -230,26 +207,27 @@ class ClassDefinition(_FrozenBaseModel):
         ),
     )
 
+    def get_content_properties(self) -> frozenset[ContentPropertyName]:
+        """Returns a set of the content properties."""
+        return frozenset(self.content.get("properties", {}))
```
[the remark this answers is not in the document]
Comment (KerstenBreuer) [reply src/schemapack/_internals/spec/schemapack.py:212]: ContentSchema is not anymore a pydantic model but a type alias for a FrozenDict. Thus it cannot contain the method anymore. The issue with caching was I think that cached properties are included in serialization by pydantic.

**Comment by KerstenBreuer** [comment]
Regarding the why not the https://github.com/Marco-Sulla/python-frozendict package question:

I am not sure whether I can recall all the reasons (we should have written an ADR) but some of them where:
- based on C not python:
  - would be more difficult to fix for us, which is very important since it is only a very small community
  - adds a compiled dependency
  - integrates with cpython and thus requires frequent changes to support multiple python versions, currently, only up to python3.10 is testes and this directory suggests that no newer versions are supported: https://github.com/Marco-Sulla/python-frozendict/tree/master/src/frozendict/c_src
  - not sure about how to attach pydantic support
- it is GPL licensed and we are using Apache 2.0
- in addition to the above, regarding the deep freeze:
  - No (at least documented way) of overwriting the default converter for dict (just for registering new ones)
  - as far as I understand it chooses the converter based on concrete classes and not generics such as Mapping (with articfreeze you can do that)

I guess we can find workarounds for these issues.

Another alternative would be (if you prefer) to simply copy the relevant source code from arcticfreeze into this repo. It is just a couple of lines. (Which anyhow demonstrates that the risks associated with this package are fairly small.) Happy to invest some time.

**Review by lkuchenb — APPROVED** [review]

**Review by lkuchenb — APPROVED** [review]
